File: lib/gamechanger/syncer.rb
# frozen_string_literal: true

module Gamechanger
  SyncResult = Struct.new(:games, :outings, :at_bats)

  # Fetches and persists games + pitcher/batter stats from the Gamechanger API.
  #
  # Extracted from CLI so sync logic can be unit-tested and reused
  # without Thor internals. Errors propagate to the caller.
  #
  # Usage:
  #   Syncer.new(config, storage).run(force: true)
  #   Syncer.new(config, storage).run(force: true, refetch_final: true)
  class Syncer
    def initialize(config, storage)
      @config  = config
      @storage = storage
    end

    # Sync all past games and their pitcher/batter stats into storage.
    #
    # The two flags are deliberately separate because they invalidate different
    # things. A final game's boxscore is immutable, so re-downloading it costs an
    # HTTP round-trip plus a Client::RATE_LIMIT_SLEEP for zero new data; only
    # non-final games actually need re-fetching on a routine sync.
    #
    # @param force [Boolean] drop cached data for non-final games first, so their
    #   rows are re-fetched from the API on this run
    # @param refetch_final [Boolean] also re-download boxscores for games already
    #   cached as final. Escape hatch for a corrupted or partially-parsed cache;
    #   costs one request + one rate-limit sleep per final game.
    # @raise [ConfigError]   if team_id or team_slug is not configured
    # @raise [AuthError]     on invalid credentials
    # @raise [NetworkError]  on connection failures or HTTP errors
    # @raise [APIShapeError] if the API response shape is unexpected
    def run(force: false, refetch_final: false)
      @storage.clear_non_final if force

      team_id   = @config.team_id
      team_slug = @config.team_slug

      if team_id.nil? || team_id.to_s.empty?
        raise ConfigError, 'No team_id configured. Run `gamechanger setup` again.'
      end

      if team_slug.nil? || team_slug.to_s.empty?
        raise ConfigError,
              'No team_slug configured. Run `gamechanger setup` again. ' \
              'Or add team_slug to ~/.gamechanger/config.yml (the short ID from your team URL).'
      end

      client     = Client.new(config: @config)
      raw_games  = client.games(team_id: team_id)
      games_list = extract_games(raw_games)
      today      = Date.today.to_s
      result     = SyncResult.new(0, 0, 0)

      games_list.each do |game|
        parsed = parse_game(game)
        next if parsed[:status] == 'canceled'
        next if parsed[:game_date].nil? || parsed[:game_date] > today

        @storage.upsert_game(parsed)

        cached = @storage.find_game(parsed[:game_id])
        next if !refetch_final && already_synced_final?(cached)

        sleep Client::RATE_LIMIT_SLEEP
        begin
          raw_boxscore = client.game_pitcher_stats(game_id: parsed[:game_id])
        rescue NetworkError => e
          # Boxscore not available yet (404) — skip this game's stats but keep
          # the schedule row so the brief can still show it as upcoming/scheduled.
          next if e.message.include?('404')

          raise
        end
        stats        = BoxscoreParser.new(raw_boxscore, team_slug: team_slug).pitcher_stats
        @storage.upsert_pitcher_stats(game_id: parsed[:game_id], stats: stats)

        batter_parser = BatterStatsParser.new(raw_boxscore, team_slug: team_slug)
        batter_stats  = batter_parser.batter_stats
        @storage.upsert_batter_stats(game_id: parsed[:game_id], stats: batter_stats) if batter_stats.any?

        fielding_stints = batter_parser.fielding_stints
        @storage.upsert_fielding_positions(game_id: parsed[:game_id], stints: fielding_stints) if fielding_stints.any?

        if stats.any?
          @storage.upsert_game(parsed.merge(status: 'final'))
          result.games   += 1
          result.outings += stats.length
          result.at_bats += batter_stats.sum { |s| s[:at_bats].to_i }
        end
      end

      result
    end

    private

    # A game is safe to skip only if it is cached as final *and* its boxscore was
    # actually parsed into stats. Status alone is not enough: the schedule feed can
    # report a game as completed before its boxscore exists, in which case we upsert
    # a 'final' row and the 404 branch below leaves it statless. Skipping on status
    # alone would strand those games with no stats forever.
    def already_synced_final?(cached)
      return false if cached.nil?

      cached['status'] == 'final' && @storage.pitcher_stats?(cached['game_id'])
    end

    # Confirmed from /teams/{uuid}/schedule?fetch_place_details=true response.
    # Returns a bare array of {event:, pregame_data:} objects (practices and games mixed).
    # Filters to event_type=="game" only.
    def extract_games(response)
      list = if response.is_a?(Array)
               response
             elsif response.is_a?(Hash)
               response['schedule'] || response['events'] || response['data'] || []
             else
               raise APIShapeError, "Unexpected schedule response shape: #{response.class}"
             end

      list.select { |item| item.dig('event', 'event_type') == 'game' }
    end

    # Confirmed field names from /teams/{uuid}/schedule response.
    # Each item is {event: {...}, pregame_data: {...}}.
    def parse_game(raw)
      event   = raw['event']
      pregame = raw['pregame_data']

      game_date = event.dig('start', 'datetime')&.split('T')&.first ||
                  event.dig('start', 'date')

      {
        game_id:   event['id'],
        game_date: game_date,
        opponent:  pregame&.dig('opponent_name') || event['title'],
        home_away: pregame&.dig('home_away'),
        status:    normalize_status(event['status'])
      }
    end

    def normalize_status(raw_status)
      case raw_status.to_s.downcase
      when /completed|final|ended/ then 'final'
      when /progress|live|active/  then 'in_progress'
      when /scheduled|upcoming/    then 'scheduled'
      else raw_status.to_s.downcase
      end
    end
  end
end

File: internal/sync/syncer.go
// Package sync fetches schedule + boxscore data from the Gamechanger API
// and persists it into store.Store. Port of Gamechanger::Syncer.
package sync

import (
	"context"
	"database/sql"
	"errors"
	"strings"
	"time"

	"github.com/joshrpowell/gamechanger-cli/internal/client"
	"github.com/joshrpowell/gamechanger-cli/internal/config"
	"github.com/joshrpowell/gamechanger-cli/internal/gcerr"
	"github.com/joshrpowell/gamechanger-cli/internal/parser"
	"github.com/joshrpowell/gamechanger-cli/internal/store"
)

// Syncer wires together client + store + parsers for a single refresh run.
type Syncer struct {
	Config *config.Config
	Client *client.Client
	Store  *store.Store

	// RateLimit is the pause between game-level requests. Zero ⇒ use the
	// client's default. Tests override to zero to keep them fast.
	RateLimit time.Duration

	// Now is a clock override for tests. Zero ⇒ time.Now().UTC().
	Now func() time.Time
}

func (s *Syncer) now() time.Time {
	if s.Now != nil {
		return s.Now()
	}
	return time.Now().UTC()
}

func (s *Syncer) rateLimit() time.Duration {
	if s.RateLimit > 0 {
		return s.RateLimit
	}
	return client.RateLimitSleep
}

// Run pulls the schedule and any unfinalized stats. force=true clears
// non-final games before syncing, forcing a re-fetch.
func (s *Syncer) Run(ctx context.Context, force bool) (store.SyncResult, error) {
	result := store.SyncResult{}

	if force {
		if err := s.Store.ClearNonFinal(ctx); err != nil {
			return result, err
		}
	}

	teamID := s.Config.TeamID
	teamSlug := s.Config.TeamSlug
	if strings.TrimSpace(teamID) == "" {
		return result, gcerr.Configf("no team_id configured. Run `gamechanger setup` again")
	}
	if strings.TrimSpace(teamSlug) == "" {
		return result, gcerr.Configf(
			"no team_slug configured. Run `gamechanger setup` again. " +
				"Or add team_slug to ~/.gamechanger/config.yml (the short ID from your team URL)")
	}

	raw, err := s.Client.Schedule(ctx, teamID)
	if err != nil {
		return result, err
	}
	scheduleItems, err := extractGames(raw)
	if err != nil {
		return result, err
	}

	today := s.now().Format("2006-01-02")
	for _, item := range scheduleItems {
		parsed, ok := parseGame(item)
		if !ok {
			continue
		}
		if parsed.status == "canceled" {
			continue
		}
		if parsed.gameDate == "" || parsed.gameDate > today {
			continue
		}

		if err := s.Store.UpsertGame(ctx, parsed.toStoreGame()); err != nil {
			return result, err
		}

		// Skip stats fetch when game is already final in cache and we
		// are not forcing a re-sync.
		cachedStatus, err := s.Store.GameStatus(ctx, parsed.gameID)
		if err != nil {
			return result, err
		}
		if cachedStatus == "final" && !force {
			continue
		}

		// Rate-limit between game-level requests.
		select {
		case <-ctx.Done():
			return result, gcerr.Networkf("context cancelled during sync")
		case <-time.After(s.rateLimit()):
		}

		raw, err := s.Client.Boxscore(ctx, parsed.gameID)
		if err != nil {
			if errors.Is(err, client.ErrBoxscoreNotFound) {
				// Canceled, postponed, or never-played game. Keep the
				// schedule row in place but skip stats for this one.
				continue
			}
			return result, err
		}
		pitcherStats, err := parser.PitcherStats(raw, teamSlug)
		if err != nil {
			return result, err
		}
		if err := s.Store.UpsertPitcherStats(ctx, parsed.gameID, pitcherStats); err != nil {
			return result, err
		}

		batterStats, err := parser.BatterStats(raw, teamSlug)
		if err != nil {
			return result, err
		}
		if len(batterStats) > 0 {
			if err := s.Store.UpsertBatterStats(ctx, parsed.gameID, batterStats); err != nil {
				return result, err
			}
		}

		if len(pitcherStats) > 0 {
			finalGame := parsed.toStoreGame()
			finalGame.Status = sql.NullString{String: "final", Valid: true}
			if err := s.Store.UpsertGame(ctx, finalGame); err != nil {
				return result, err
			}
			result.Games++
			result.Outings += len(pitcherStats)
			for _, b := range batterStats {
				result.AtBats += b.AtBats
			}
		}
	}

	return result, nil
}

// parsedGame mirrors the keys Ruby's Syncer#parse_game produces. Internal
// to this package so the store.Game shape stays uncontaminated.
type parsedGame struct {
	gameID   string
	gameDate string
	opponent string
	homeAway string
	status   string
}

func (p parsedGame) toStoreGame() store.Game {
	return store.Game{
		GameID:   p.gameID,
		GameDate: p.gameDate,
		Opponent: nullableString(p.opponent),
		HomeAway: nullableString(p.homeAway),
		Status:   nullableString(p.status),
	}
}

// extractGames flattens the API's schedule response into a list of
// {event, pregame_data} maps, filtered to event_type == "game".
func extractGames(raw any) ([]map[string]any, error) {
	var items []any
	switch v := raw.(type) {
	case []any:
		items = v
	case map[string]any:
		for _, key := range []string{"schedule", "events", "data"} {
			if list, ok := v[key].([]any); ok {
				items = list
				break
			}
		}
		if items == nil {
			return nil, gcerr.APIShapef("unexpected schedule response shape: %T", raw)
		}
	default:
		return nil, gcerr.APIShapef("unexpected schedule response shape: %T", raw)
	}

	out := make([]map[string]any, 0, len(items))
	for _, it := range items {
		m, ok := it.(map[string]any)
		if !ok {
			continue
		}
		event, _ := m["event"].(map[string]any)
		if event == nil {
			continue
		}
		if et, _ := event["event_type"].(string); et != "game" {
			continue
		}
		out = append(out, m)
	}
	return out, nil
}

func parseGame(item map[string]any) (parsedGame, bool) {
	event, _ := item["event"].(map[string]any)
	pregame, _ := item["pregame_data"].(map[string]any)
	if event == nil {
		return parsedGame{}, false
	}

	gameID, _ := event["id"].(string)
	if gameID == "" {
		return parsedGame{}, false
	}

	// Date can come from start.datetime (split on 'T') or start.date.
	gameDate := ""
	if start, ok := event["start"].(map[string]any); ok {
		if dt, _ := start["datetime"].(string); dt != "" {
			if idx := strings.IndexByte(dt, 'T'); idx > 0 {
				gameDate = dt[:idx]
			} else {
				gameDate = dt
			}
		} else if d, _ := start["date"].(string); d != "" {
			gameDate = d
		}
	}

	// opponent: pregame.opponent_name OR event.title
	opponent := ""
	if pregame != nil {
		if s, _ := pregame["opponent_name"].(string); s != "" {
			opponent = s
		}
	}
	if opponent == "" {
		if t, _ := event["title"].(string); t != "" {
			opponent = t
		}
	}

	homeAway := ""
	if pregame != nil {
		homeAway, _ = pregame["home_away"].(string)
	}

	rawStatus, _ := event["status"].(string)
	return parsedGame{
		gameID:   gameID,
		gameDate: gameDate,
		opponent: opponent,
		homeAway: homeAway,
		status:   normalizeStatus(rawStatus),
	}, true
}

func normalizeStatus(raw string) string {
	s := strings.ToLower(strings.TrimSpace(raw))
	switch {
	case strings.Contains(s, "completed"), strings.Contains(s, "final"), strings.Contains(s, "ended"):
		return "final"
	case strings.Contains(s, "progress"), strings.Contains(s, "live"), strings.Contains(s, "active"):
		return "in_progress"
	case strings.Contains(s, "scheduled"), strings.Contains(s, "upcoming"):
		return "scheduled"
	default:
		return s
	}
}

func nullableString(s string) sql.NullString {
	if s == "" {
		return sql.NullString{}
	}
	return sql.NullString{String: s, Valid: true}
}
